### PR TITLE
Refactor ModRef out as a type.

### PR DESCRIPTION
**Pull request by associahedron**
This is in preparation for moving to a monadic style.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs [hunk 1]
```diff
@@ -193,94 +199,94 @@ generateSrcFromLf env = noLoc mod
...
             , ideclImplicit = False
             , ideclQualified = True
             , ideclAs = Nothing
             , ideclHiding = Nothing
             } :: LImportDecl GhcPs
-        | (isStable, refUnitId, modRef) <- modRefs
+        | ModRef{..} <- Set.toList modRefs
          -- don’t import ourselves
-        , not (modRef == lfModName && unitId == unitId)
+        , not (modRefModule == lfModName && modRefUnitId == unitId)
```
Comment (associahedron) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs]: I *think* this is an unrelated bugfix.

Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs]: This really should be in HLint :slightly_smiling_face:

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs [hunk 2]
```diff
@@ -193,94 +199,94 @@ generateSrcFromLf env = noLoc mod
...
-            , ideclPkgQual =
-              if isStable
-                then Nothing -- we don’t do package qualified imports for modules that should come from the current SDK.
-                else Just $ StringLiteral NoSourceText $ mkFastString $
-                  -- Package qualified imports for the current package need to use "this" instead
-                  -- of the package id.
-                  if refUnitId == unitId then "this" else fst (splitUnitId $ unitIdString refUnitId)
+                  noLoc $ mkModuleName $ T.unpack $ LF.moduleNameString modRefModule
+            , ideclPkgQual = do
+                guard modRefIsStable -- we don’t do package qualified imports
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs]:
```suggestion
                guard $ not modRefIsStable -- we don’t do package qualified imports
```

Comment (associahedron) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs]: oops

**Comment by hurryabit** [comment]
Thanks. May I ask what exactly is moving to a monadic style?

**Comment by associahedron** [comment]
> Thanks. May I ask what exactly is moving to a monadic style?

The converter in DataDependencies.hs, so we can emit module references at the same time as the conversion takes place, instead of reconstructing the logic for what gets exported twice.

**Comment by hurryabit** [comment]
> The converter in DataDependencies.hs, so we can emit module references at the same time as the conversion takes place, instead of reconstructing the logic for what gets exported twice.

I understand. Makes a lot of sense. Thanks for the explanation.
